cirq-core/cirq/transformers/measurement_transformers_test.py

```diff
@@ -307,12 +307,100 @@ def test_sympy_control():
+    circuit = cirq.Circuit(
+        cirq.measure(
+            q0,
+            q1,
+            key='a',
+            confusion_map={
+                (0, 1): np.array(
+                    [
+                        [0.7, 0.1, 0.1, 0.1],
+                        [0.1, 0.6, 0.1, 0.2],
+                        [0.2, 0.2, 0.5, 0.1],
+                        [0.0, 0.0, 1.0, 0.0],
+                    ]
+                )
+            },
+        ),
+        cirq.X(q2).with_classical_controls('a'),
+        cirq.measure(q2, key='b'),
+    )
+    deferred = cirq.defer_measurements(circuit)
+    sim = cirq.DensityMatrixSimulator()
+    result = sim.sample(deferred, repetitions=10_000)
+
+    # The initial state is zero, so the first measurement will confuse by the first line in the
+    # map, giving 7000 0's, 1000 1's, 1000 2's, and 1000 3's, for a sum of 6000 on average.
+    assert 5_600 <= np.sum(result['a']) <= 6_400
+
+    # The measurement will be non-zero 3000 times on average.
+    assert 2_600 <= np.sum(result['b']) <= 3_400
+
+    # Try a deterministic one: initial state is 3, which the confusion map sends to 2 with p=1.
+    deferred.insert(0, cirq.X.on_each(q0, q1))
+    result = sim.sample(deferred, repetitions=100)
+    assert np.sum(result['a']) == 200
+    assert np.sum(result['b']) == 100
+
+
+def test_confusion_map_errors():
+    q0, q1 = cirq.LineQubit.range(2)
+    circuit = cirq.Circuit(
+        cirq.measure(q0, key='a', confusion_map={(0,): np.array([1])}),
+        cirq.X(q1).with_classical_controls('a'),
+    )
+    with pytest.raises(ValueError, match='map must be 2D'):
+        _ = cirq.defer_measurements(circuit)
+    circuit = cirq.Circuit(
+        cirq.measure(q0, key='a', confusion_map={(0,): np.array([[0.7, 0.3]])}),
+        cirq.X(q1).with_classical_controls('a'),
+    )
+    with pytest.raises(ValueError, match='map must be square'):
+        _ = cirq.defer_measurements(circuit)
+    circuit = cirq.Circuit(
+        cirq.measure(
+            q0,
+            key='a',
+            confusion_map={(0,): np.array([[0.7, 0.1, 0.2], [0.1, 0.6, 0.3], [0.2, 0.2, 0.6]])},
+        ),
+        cirq.X(q1).with_classical_controls('a'),
+    )
+    with pytest.raises(ValueError, match='size does not match'):
+        _ = cirq.defer_measurements(circuit)
+    circuit = cirq.Circuit(
+        cirq.measure(q0, key='a', confusion_map={(0,): np.array([[-1, 2], [0, 1]])}),
+        cirq.X(q1).with_classical_controls('a'),
+    )
+    with pytest.raises(ValueError, match='negative probabilities'):
+        _ = cirq.defer_measurements(circuit)
+    circuit = cirq.Circuit(
+        cirq.measure(q0, key='a', confusion_map={(0,): np.array([[0.3, 0.3], [0.3, 0.3]])}),
         cirq.X(q1).with_classical_controls('a'),
     )
-    with pytest.raises(
-        NotImplementedError, match='Deferring confused measurement is not implemented'
-    ):
+    with pytest.raises(ValueError, match='invalid probabilities'):
```

Comment:
Hmm, compares to what? Confusion is something that happens to the cbits, so there's no way of getting a density matrix on the original circuit that has anything to do with the confusion matrix, and thus no way to compare the deferred circuit's DM to the original.

I guess I _could_ change the measurement to a dephase on the original circuit and then check that the partial trace on the _original_ qubit of the deferred circuit is equivalent to that. Though that only checks that the confusion matrix does _not_ affect the original qubit. Probably that's worthwhile to verify, but I get the feeling you had something more valuable in mind.

Or did you just mean to compare directly to the density matrix that we expect in some simple case?

Reply:
What will the density matrix sim do when the confused cbit controls a unitary gate? I thought it would produce the density matrix that is a convex combination of the results of executions with and without the controlled unitary gate where convex combination coefficients would reflect the confusion matrix. Is this not the case?

Reply:
When DM simulator encounters a measurement gate it collapses. If the user doesn't want this, they can use the `dephase_measurements` transformer that changes measurements to dephase ops, and run the simulation on the transformed circuit instead. But that fails in the presence of classical controls. So the deferred measurements transformer essentially _is_ what you use to get the density matrix from a circuit that has classical controls. Thus there's nothing else to compare against.

I was trying at one point to have the DM simulator do this transformation automatically during simulation, but it was decided that we should have it be an explicit AOT transformation instead, which is where the deferred measurement transformer came from. If we'd _like_ to have the DM simulator do this automatically, it could be an interesting starter project for someone. Then _that_ could have tests that compare against the DM generated by _this_.

But anyway, do you see anything left to do here? This is the final remaining open issue.

Reply:
I'll go ahead and write up a test that compares the DM to a known value....